### PR TITLE
Admission Controller POC

### PR DESCRIPTION
**Pull request by cezarybednarz**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,18 +1,19 @@
-REPO        		?= mmazurczyk
+REPO        		?= cezarybednarz
```
Comment (zimnx) [inline Makefile]: I added repos for all these three images, you guys should have access to push images there:
https://hub.docker.com/r/scyllazimnx/scylla-operator-autoscaler-admission-controller
https://hub.docker.com/r/scyllazimnx/scylla-operator-autoscaler-recommender
https://hub.docker.com/r/scyllazimnx/scylla-operator-autoscaler-updater
Please use these here

File: Makefile [hunk 2]
```diff
@@ -1,18 +1,19 @@
...
 
 .EXPORT_ALL_VARIABLES:
-DOCKER_BUILDKIT         := 1
-GOVERSION               := $(shell go version)
-GOPATH                  := $(shell go env GOPATH)
-KUBEBUILDER_ASSETS      := $(GOPATH)/bin
-PATH                    := $(GOPATH)/bin:$(PATH):
-RECOMMENDER_IMG         := $(REPO)/$(IMG_PREFIX)recommender
-UPDATER_IMG         := $(REPO)/$(IMG_PREFIX)updater
+DOCKER_BUILDKIT        		 	:= 1
```
Comment (zimnx) [inline Makefile]: It's not used by anything

File: cmd/admission-controller/admission_controller.go [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+				logger.Error(ctx, "Unable to set up overall controller manager", err)
+				os.Exit(1)
+			}
+
+			// setup webhooks
+			logger.Info(ctx, "Setting up webhook server")
+			webhookServer := mgr.GetWebhookServer()
+
+			logger.Info(ctx, "Registering webhooks to the webhook server")
+			webhookServer.Register("/mutate-scylla-scylladb-com-v1-scyllacluster", &webhook.Admission{Handler: &recommendationApplier{Client: mgr.GetClient(), logger: logger}})
```
Comment (zimnx) [inline cmd/admission-controller/admission_controller.go]: line is too long

File: cmd/admission-controller/main.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func main() {
+	ctx := log.WithNewTraceID(context.Background())
+	atom := zap.NewAtomicLevelAt(zapcore.InfoLevel)
+	logger, _ := log.NewProduction(log.Config{
+		Level: atom,
+	})
+
+	var rootCmd = &cobra.Command{}
```
Comment (zimnx) [inline cmd/admission-controller/main.go]: Let's drop `cobra` from this binary. There're no config files nor parameters so far.

File: cmd/admission-controller/mutatingwebhook.go [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+	return nil
+}
+
+func (ra *recommendationApplier) Handle(ctx context.Context, req admission.Request) admission.Response {
+	cluster := &scyllav1.ScyllaCluster{}
+
+	err := ra.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]:
```suggestion
	if err := ra.decoder.Decode(req, cluster); err != nil {
		return admission.Errored(http.StatusBadRequest, err)
	}
```

Comment (zimnx) [reply cmd/admission-controller/mutatingwebhook.go]: Same applies below

File: cmd/admission-controller/mutatingwebhook.go [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+	c, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: scheme})
+	if err != nil {
+		return fmt.Errorf("Failed to create a client: %s", err)
+	}
+
+	scas := &v1alpha1.ScyllaClusterAutoscalerList{}
+	err = c.List(ctx, scas)
+	if err != nil {
+		return fmt.Errorf("Failed to get SCAs: %s", err)
+	}
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]:
```suggestion
	if err = c.List(ctx, scas); err != nil {
		return fmt.Errorf("Failed to get SCAs: %s", err)
	}
```

File: cmd/admission-controller/mutatingwebhook.go [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+	return nil
+}
+
+func mutateCluster(ctx context.Context, logger log.Logger, cluster *scyllav1.ScyllaCluster) error {
+	logger.Info(ctx, "Starting mutation of ScyllaCluster")
+
+	c, err := client.New(config.GetConfigOrDie(), client.Options{Scheme: scheme})
+	if err != nil {
+		return fmt.Errorf("Failed to create a client: %s", err)
+	}
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: Please create client on initialization, not on every request

File: cmd/admission-controller/mutatingwebhook.go [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+
+	scas := &v1alpha1.ScyllaClusterAutoscalerList{}
+	err = c.List(ctx, scas)
+	if err != nil {
+		return fmt.Errorf("Failed to get SCAs: %s", err)
+	}
+
+	logger.Debug(ctx, "SCAs fetched", "num", len(scas.Items))
+
+	for idx := range scas.Items {
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: Shouldn't this list be filetered to find SCA pointing to cluster under mutation?

File: cmd/admission-controller/mutatingwebhook.go [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+		dcRecs := getDataCenterRecommendations(sca)
+		if dcRecs == nil {
+			logger.Debug(ctx, "No recommendations for cluster", "cluster", cluster.Name)
+			continue
+		}
+
+		dataCenterName := cluster.Spec.Datacenter.Name
+
+		if cluster.Spec.Datacenter.Name != dataCenterName {
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: This is always false

File: cmd/admission-controller/mutatingwebhook.go [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+func getRackRecommendations(dataCenterName string,
+	dcRecs []v1alpha1.DataCenterRecommendations) []v1alpha1.RackRecommendations {
+	for idx := range dcRecs {
+		if dcRecs[idx].Name == dataCenterName {
+			if dcRecs[idx].RackRecommendations == nil {
+				return nil
+			}
+			return dcRecs[idx].RackRecommendations
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]:
```suggestion
			return dcRecs[idx].RackRecommendations
```

File: cmd/admission-controller/mutatingwebhook.go [hunk 11]
```diff
@@ -0,0 +1,157 @@
...
+
+	logger log.Logger
+}
+
+func getDataCenterRecommendations(sca *v1alpha1.ScyllaClusterAutoscaler) []v1alpha1.DataCenterRecommendations {
+	if sca.Status.Recommendations == nil {
+		return nil
+	}
+	dcRecs := sca.Status.Recommendations.DataCenterRecommendations
+	if dcRecs == nil || len(dcRecs) == 0 {
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: len(nil) == 0, so you can simply check length

File: cmd/admission-controller/mutatingwebhook.go [hunk 12]
```diff
@@ -0,0 +1,157 @@
...
+				continue
+			}
+
+			rack := findRack(rackRec.Name, cluster.Spec.Datacenter.Racks)
+			if rack == nil {
+				logger.Debug(ctx, "Could not find rack matching recommendation", "rack", rackRec.Name)
+				continue
+			}
+
+			if rackRec.Name != rack.Name {
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: first you're searching for rack in findRack(), and then you double check if it's the rack you were looking for. Trust yourself :D

File: cmd/admission-controller/mutatingwebhook.go [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+
+	err := ra.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	err = mutateCluster(ctx, ra.logger, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]:
```suggestion
	if err := mutateCluster(ctx, ra.logger, cluster); err != nil {
		return admission.Errored(http.StatusInternalServerError, err)
	}
```

File: config/certmanager/certificate.yaml [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+  namespace: system
+spec:
+  # $(SERVICE_NAME) and $(SERVICE_NAMESPACE) will be substituted by kustomize
+  dnsNames:
+    - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc
+    - $(SERVICE_NAME).$(SERVICE_NAMESPACE).svc.cluster.local
+  issuerRef:
+    kind: Issuer
+    name: selfsigned-issuer
+  secretName: webhook-server-cert # this secret will not be prefixed, since it's not managed by kustomize
```
Comment (zimnx) [inline config/certmanager/certificate.yaml]: missing empty line

File: config/certmanager/kustomization.yaml [hunk 15]
```diff
@@ -0,0 +1,5 @@
+resources:
+  - certificate.yaml
+
+configurations:
+  - kustomizeconfig.yaml
```
Comment (zimnx) [inline config/certmanager/kustomization.yaml]: missing empty line

File: config/certmanager/kustomizeconfig.yaml [hunk 16]
```diff
@@ -0,0 +1,16 @@
...
+        group: cert-manager.io
+        path: spec/issuerRef/name
+
+varReference:
+  - kind: Certificate
+    group: cert-manager.io
+    path: spec/commonName
+  - kind: Certificate
+    group: cert-manager.io
+    path: spec/dnsNames
```
Comment (zimnx) [inline config/certmanager/kustomizeconfig.yaml]: missing empty line

**Review by zimnx — CHANGES_REQUESTED** [review]

**Comment by zimnx** [comment]
It doesn't work, update API call is missing. Please test it first, and resolve conflicts.

File: cmd/admission-controller/mutatingwebhook.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+func getDataCenterRecommendations(sca *v1alpha1.ScyllaClusterAutoscaler) []v1alpha1.DataCenterRecommendations {
+	if sca.Status.Recommendations == nil {
+		return nil
+	}
+	dcRecs := sca.Status.Recommendations.DataCenterRecommendations
+	if len(dcRecs) == 0 {
+		return nil
+	}
+	return dcRecs
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: nil slice and zero len slice should be treated the same, you can simply return `sca.Status.Recommendations.DataCenterRecommendations`

File: cmd/admission-controller/mutatingwebhook.go (hunk 8 above)
Comment (zimnx) [reply cmd/admission-controller/mutatingwebhook.go]: This comment is not resolved.

File: cmd/admission-controller/mutatingwebhook.go [hunk 18]
```diff
@@ -0,0 +1,150 @@
...
+		return fmt.Errorf("Failed to get SCAs: %s", err)
+	}
+
+	logger.Debug(ctx, "SCAs fetched", "num", len(scas.Items))
+
+	for idx := range scas.Items {
+		sca := &scas.Items[idx]
+
+		if *sca.Spec.UpdatePolicy.UpdateMode == v1alpha1.UpdateModeOff {
+			logger.Debug(ctx, "Autoscaler has 'off' scaling policy, skipping", "autoscaler", sca.ObjectMeta.Name)
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]:
```suggestion
			logger.Debug(ctx, "Autoscaler has 'off' scaling policy, skipping", "autoscaler", sca.Name)
```

File: cmd/admission-controller/mutatingwebhook.go [hunk 19]
```diff
@@ -0,0 +1,150 @@
...
+
+		targetRef := sca.Spec.TargetRef
+		referencedCluster := &scyllav1.ScyllaCluster{}
+		var err = c.Get(ctx, client.ObjectKey{
+			Namespace: targetRef.Namespace,
+			Name:      targetRef.Name,
+		}, cluster)
+		if err != nil {
+			return fmt.Errorf("Fetch referenced ScyllaCluster: %s", err)
+		}
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: It's pointless to fetch it here. You have cluster under mutation in function parameters, you can simply compare targetRef name and namespace with it without fetching.

File: cmd/admission-controller/mutatingwebhook.go [hunk 20]
```diff
@@ -0,0 +1,150 @@
...
+
+		if referencedCluster != cluster {
+			logger.Debug(ctx, "SCA not pointing to cluster under mutation")
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: return is missing

File: cmd/admission-controller/mutatingwebhook.go [hunk 21]
```diff
@@ -0,0 +1,150 @@
...
+		referencedCluster := &scyllav1.ScyllaCluster{}
+		var err = c.Get(ctx, client.ObjectKey{
+			Namespace: targetRef.Namespace,
+			Name:      targetRef.Name,
+		}, cluster)
+		if err != nil {
+			return fmt.Errorf("Fetch referenced ScyllaCluster: %s", err)
+		}
+
+		if referencedCluster != cluster {
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: This isn't going to work, pointers nor deeper value won't be equal. You're only interested in identifier which is UID or pair of namespace and name.

File: cmd/admission-controller/mutatingwebhook.go [hunk 22]
```diff
@@ -0,0 +1,150 @@
...
+			logger.Debug(ctx, "SCA not pointing to cluster under mutation")
+		}
+
+		logger.Debug(ctx, "Cluster has 'Initial' or 'Auto' scaling policy")
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: Why not log which one is used?

File: cmd/admission-controller/mutatingwebhook.go [hunk 23]
```diff
@@ -0,0 +1,150 @@
...
+			rackRec := &rackRecs[j]
+
+			if rackRec.Members == nil {
+				logger.Debug(ctx, "No members recommendation for rack", "rack", rackRec.Name)
+				continue
+			}
+
+			rack := findRack(rackRec.Name, cluster.Spec.Datacenter.Racks)
+			if rack == nil {
+				logger.Debug(ctx, "Could not find rack matching recommendation", "rack", rackRec.Name)
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go]: This should be logged as error, not debug.

File: cmd/admission-controller/mutatingwebhook.go [hunk 24]
```diff
@@ -0,0 +1,150 @@
...
+				continue
+			}
+
+			rack.Members = rackRec.Members.Target
+
```
Comment (zimnx) [inline cmd/admission-controller/mutatingwebhook.go:119]: Update is missing, you're modyfing rack in memory of admission controller process.

**Review by zimnx — CHANGES_REQUESTED** [review]
Left comments inline, please add unit tests.

File: cmd/admission-controller/validatingwebhook.go [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+const (
+	updaterServiceAccountUsername = "system:serviceaccount:scylla-operator-autoscaler-system:scylla-operator-autoscaler-updater-service-account"
+)
+
+func validateClusterChanges(ctx context.Context, logger log.Logger, cluster *scyllav1.ScyllaCluster, oldCluster *scyllav1.ScyllaCluster, scas *v1alpha1.ScyllaClusterAutoscalerList) error {
+	logger.Info(ctx, "starting validation of ScyllaCluster")
+
+	for idx := range scas.Items {
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook.go:33]: nit: you can use `for _, sca := range scas.Items`
same in inner loop

File: cmd/admission-controller/validatingwebhook.go [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+		for idr := range cluster.Spec.Datacenter.Racks {
+			rack := cluster.Spec.Datacenter.Racks[idr]
+
+			oldRack := v1.RackSpec{}
+
+			for idrOld := range oldCluster.Spec.Datacenter.Racks {
+				r := oldCluster.Spec.Datacenter.Racks[idrOld]
+				if rack.Name == r.Name {
+					oldRack = r
+					break
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook.go:58]: what if it's not found?

File: cmd/admission-controller/validatingwebhook.go [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+				}
+			}
+
+			if rack.Members != oldRack.Members {
+				return fmt.Errorf("changing members is forbidden while cluster is administered by autoscaler")
+			}
+
+			if !rack.Resources.Requests.Cpu().Equal(*oldRack.Resources.Requests.Cpu()) {
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook.go:66]: it has to check entire resources, not just requests.cpu.

File: cmd/admission-controller/validatingwebhook.go [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+type admissionValidator struct {
+	Client  client.Client
+	decoder *admission.Decoder
+
+	scyllaClient client.Client
+	logger       log.Logger
+}
+
+const (
+	updaterServiceAccountUsername = "system:serviceaccount:scylla-operator-autoscaler-system:scylla-operator-autoscaler-updater-service-account"
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook.go:27]: This should be a CLI parameter with default value

File: cmd/admission-controller/validatingwebhook_test.go [hunk 29]
```diff
@@ -0,0 +1,133 @@
...
+		name       string
+		cluster    *v1.ScyllaCluster
+		oldCluster *v1.ScyllaCluster
+		scas       *v1alpha1.ScyllaClusterAutoscalerList
+		allowed    bool
+	}{
+		{
+			name:       "unchanged cluster",
+			cluster:    basicCluster,
+			oldCluster: oldBasicCluster,
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook_test.go:82]: if it's the same, why you're using different variable? I know it's the same, but it's odd.

File: cmd/admission-controller/validatingwebhook_test.go [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+			scas:       basicScas,
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook_test.go:83]: what is basic? name should describe what's so special about it, wheter it's auto/off mode, whether it contains single rack node increasing recomendation etc

File: cmd/admission-controller/validatingwebhook_test.go [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+
+	tests := []struct {
+		name       string
+		cluster    *v1.ScyllaCluster
+		oldCluster *v1.ScyllaCluster
+		scas       *v1alpha1.ScyllaClusterAutoscalerList
+		allowed    bool
+	}{
+		{
+			name:       "unchanged cluster",
```
Comment (zimnx) [inline cmd/admission-controller/validatingwebhook_test.go:80]: naming of tests could be improved to describe expected behavior. "allow empty update", "deny changing CPU limits" etc

File: config/admission-controller/admission_controller.yaml [hunk 32]
```diff
@@ -0,0 +1,34 @@
...
+      containers:
+        - command:
+            - /admission-controller
+          args:
+            - admission-controller
+          image: admission-controller:latest
+          imagePullPolicy: Always
+          name: admission-controller
+          resources:
+            limits:
```
Comment (zimnx) [inline config/admission-controller/admission_controller.yaml:28]: remove the limit

**Review by zimnx — COMMENTED** [review]
good job, some test refactorig is needed

File: cmd/admission-controller/validatingwebhook.go (hunk 27 above)
Comment (cezarybednarz) [reply cmd/admission-controller/validatingwebhook.go:66]: I think we care only about requests.cpu because it's the only value scaled by autoscaler

Comment (zimnx) [reply cmd/admission-controller/validatingwebhook.go:66]: and limits.cpu (https://github.com/scylladb/scylla-operator-autoscaler/pull/19/files#diff-36d043d1b3fe2d0a248f67a2c154427e46a67dfc966a1343afaf28a3348c7c61R231)

**Comment by cezarybednarz** [comment]
reopen

**Pull request closed without merging**